### DISCUSSION TITLE
Bug Proofing State Transition Implementations

### DISCUSSION THREAD

**Discussion by VinayaSathyanarayana** [statelyai/xstate#1544] [Ideas]
I have seen bugs being introduced multiple times when the developer

* forgot to assign new state
* new state is assigned irrespective of whether the transition was successful or not
* states for error conditions are not designed at all or missed out

Examples could be database update fails, then the state has to reflect that correctly

Should we not have explicit mechanism to help make the state diagram robust
Current_State --> Transition --> next.state_If_Transition is Success, next.state_If_Transition_Fails

There could be multiple next.states based on the error code

How do we enhance the xstate library to do this?

**Comment by davidkpiano** [comment]
How would you define a transition as "failing"?

In state machines, everything is a "signal" (i.e., "event"). So an error is an event just like any other event.

Would you want some strict mode to ensure that e.g., `onError` is defined for all invocations?
